### PR TITLE
Poll UPS to check for deleted variants

### PR DESCRIPTION
**Pull request by darahayes**
Related Ticket: https://issues.jboss.org/browse/AEROGEAR-2647

## Description

This PR implements a mechanism where every 10 seconds, the service polls UPS to build a list of all variants. It then compares those variants against the UPS secrets created in K8s to see if there are any client configs that reference a variant that no longer exists in UPS. If we find a client UPS config that references a variant that does not exist, we assume the variant has been deleted and we trigger an unbind by deleting the servicebinding relating to that particular variant.

One issue I have found is that when we notice a variant has been deleted in UPS, we delete the servicebinding, this results in the bind secret being deleted which results in the `handleDelete` function being triggered which makes a call to UPS to delete the variant but the variant has already been deleted. Not sure what we should do here...

### Some additional notes

* I've also done some minor refactoring in the pushClient.go
* My editor has some automatic `gofmt` thing so you may see some stylistic changes in places.
* `main.go` is getting really bloated and readability is suffering big time. This should definitely be resolved in another PR but for now, I've tried to make some minor improvements by bringing the `main()`, `startWatchLoop()` and `startPollingUPS()` functions to the top.

### To Do

- [x] some of the function and variable names are probably not the best and could be renamed
- [x] I'm trying to figure out how to make the code simpler to understand
- [x] need to refactor the `pushClientOrDie` function to not die if an error occurs. 
- [x]  We can then implement the appropriate nil checks and error checks in places we use the pushClient.

@pb82 would love to get your thoughts on this.

### Setup

To make it easier to verify, I've pushed an image based off this PR to `docker.io/darahayes/ups-config-operator:master`. This means you don't need to clone the code and build locally. You can trust that the code builds based off the build in CircleCI.

* Ensure you have the latest unifiedpush-apb image with `docker pull aerogear/catalog/unifiedpush-apb`
* Provision UPS from the service catalog as normal. Do not create any bindings at this time
* In OpenShift, go into deployments/ups and replace `docker.io/aerogear/ups-config-operator:master` with `docker.io/darahayes/ups-config-operator:master`. Hit save. This should kick off a new deployment.
* Create some bindings on the UPS service for iOS/Android clients.
* You should see the associated variants in UPS and also under resoruces/secrets in OpenShift you should find a secret with a name like `ups-secret-<my-client-id>-abcjd` inside the `config` property of this secret(s) you will find the config that should be consumed by the client. There will be fields for `android` and `ios` depending on the bindings you created.

### Actions

* In the UPS console, delete a variant.
* After a few seconds, verify that the servicebinding associated with this variant has been removed.
* verify that the corresponding `ios` or `android` config has been removed from the secret in OpenShift

### CODE REVIEW

**Comment by josemigallas** [comment]
@darahayes After deleting a variant in UPS console, I still can see a secret with this info in Openshift:
```
{"CLIENT_TYPE":"Android","iosIsProduction":false,"CLIENT_ID":"binding02","googlekey":"456","projectNumber":"456"}
```

**Comment by darahayes** [comment]
@josemigallas This wouldn't have happened as a result of the code in this PR. This PR simply kicks off the unbind action where appropriate. That secret you see leftover is a bind secret which would be deleted by the unbind role in the unifiedpush-apb. 

Can you pull the latest unifiedpush-apb and start the verification steps again?

**Comment by josemigallas** [comment]
Verified, the secret `ups-secret-<my-client-id>-abcjd` has been removed.

File: cmd/server/main.go [hunk 1]
```diff
@@ -51,6 +51,9 @@ const IOSCert = "cert"
 const IOSPassPhrase = "passphrase"
 const IOSIsProduction = "isProduction"
 
+// time in seconds
+const UPSPollingInterval = 10
```
Comment (wtrocki) [inline cmd/server/main.go]: IMHO 60 seconds is enough.

**Comment by pb82** [comment]
@josemigallas I would have thought that the binding secret gets automatically removed by the ASB, but looks like thats not the case. We can look into removing that secret in the (currently unused) unbind role.

The more important things to test are:

* if there was only one variant, is the UPS secret deleted?
* if there were two secrets, is the config object inside that secret updated (one of the platforms removed)?

**Comment by pb82** [comment]
@darahayes Yes by triggering unbind the config operator will also make a delete http call to UPS even though the variant no longer exists. I don't think thats a problem and i like how we just use the regular unbind flow instead of implementing everything explicitely for this scenario.

**Comment by josemigallas** [comment]
@pb82 With one variant the secret is removed.

On the other hand, I don't see the iOS variant inside the secret's config, only Android. It is not in UPS console either. The key is valid because I can create it from the UPS console and it works but not when creating a binding. Maybe I'm missing something.

**Comment by cfoskin** [comment]
@josemigallas  if on fedora try using 

`cat Cert.p12 | base64 --wrap=0`

There is an issue with the cert output on Fedora

**Comment by cfoskin** [comment]
@pb82  We have it listed on the docs task for UPS management via the Openshift UI. not sure where that will live

File: cmd/server/main.go (hunk 1 above)
Comment (pb82) [reply cmd/server/main.go]: The loop isn't that costly (http request on the same pod, number of variants usually small). With this in mind i would rather leave it at 10 seconds for better response times.

File: cmd/server/main.go [hunk 2]
```diff
@@ -62,6 +65,210 @@ type BindingSecret struct {
...
+		log.Printf("An error occurred trying to get variants from UPS service: %v", err.Error())
+		return
+	}
+
+	for _, clientConfig := range clientConfigs {
+		found := false
+
+		for _, variant := range UPSVariants {
+			if variant.VariantID == clientConfig.VariantId {
+				found = true
```
Comment (pb82) [inline cmd/server/main.go]: We could exit the loop early here, no reason to inspect the other variants when we already found the matching one

File: cmd/server/main.go [hunk 3]
```diff
@@ -62,6 +65,210 @@ type BindingSecret struct {
...
+}
+
+// Find a service binding by its ExternalID
+func getServiceBindingNameByID(bindingId string) (string, error) {
+	// Get a list of all service bindings in the namespace and find the one with a matching ExternalID
+	// This is not very efficient and could be improved with a jsonpath query but it looks like client-go
+	// does not support jsonpath or at least I could not find any examples.
+	bindings, err := scclient.ServicecatalogV1beta1().ServiceBindings(os.Getenv(NamespaceKey)).List(metav1.ListOptions{})
+	if err != nil {
+		return "", errors.New("Error listing service bindings")
```
Comment (pb82) [inline cmd/server/main.go]: I wrote this, but maybe we should return the existing error here instead of creating a new one, hiding the real error message.

**Pull request merged**